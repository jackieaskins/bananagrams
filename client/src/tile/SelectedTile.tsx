import { useEffect, useState } from 'react';
import { useRecoilValue } from 'recoil';

import Tile from './Tile';
import { selectedTileState } from './selectedTileState';

const SelectedTile = (): JSX.Element | null => {
  const [mousePosition, setMousePosition] = useState({ x: 0, y: 0 });
  const selectedTile = useRecoilValue(selectedTileState);

  useEffect(() => {
    const onMouseMove = ({ clientX: x, clientY: y }: MouseEvent): void => {
      setMousePosition({ x, y });
    };

    document.addEventListener('mousemove', onMouseMove);
    return () => document.removeEventListener('mousemove', onMouseMove);
  }, []);

  if (!selectedTile) {
    return null;
  }

  const { x, y } = mousePosition;
  const {
    tile: { id, letter },
  } = selectedTile;

  return (
    <div
      css={{
        position: 'fixed',
        left: `${x + 5}px`,
        top: `${y + 5}px`,
        zIndex: 1000000,
      }}
    >
      <Tile id={id} letter={letter} />
    </div>
  );
};

export default SelectedTile;
