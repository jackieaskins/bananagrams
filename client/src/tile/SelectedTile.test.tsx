import { shallow } from 'enzyme';
import { useRecoilValue } from 'recoil';

import SelectedTile from './SelectedTile';

const mockSetMousePosition = jest.fn();
jest.mock('react', () => ({
  ...jest.requireActual<any>('react'),
  useEffect: jest.fn((fn) => fn()()),
  useState: jest.fn((init) => [init, mockSetMousePosition]),
}));

const mockUseRecoilValue = useRecoilValue as jest.Mock;
jest.mock('recoil', () => ({
  useRecoilValue: jest.fn(),
}));

jest.mock('./selectedTileState', () => ({
  selectedTileState: 'selectedTileState',
}));

const mockAddEventListener = jest.fn();
const mockRemoveEventListener = jest.fn();
describe('<SelectedTile />', () => {
  const renderComponent = (propOverrides = {}) =>
    shallow(<SelectedTile {...propOverrides} />);

  beforeEach(() => {
    global.document.addEventListener = mockAddEventListener;
    global.document.removeEventListener = mockRemoveEventListener;
  });

  it('does not render when no tile is selected', () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  it('renders Tile when tile is selected', () => {
    mockUseRecoilValue.mockReturnValue({
      tile: { letter: 'A', id: 'A1' },
      boardPosition: null,
    });

    expect(renderComponent()).toMatchSnapshot();
  });

  it('adds listener on mouse move', () => {
    renderComponent();

    expect(global.document.addEventListener).toHaveBeenCalledWith(
      'mousemove',
      expect.any(Function)
    );
  });

  it('removes mouse move listener on dismount', () => {
    renderComponent();

    expect(global.document.removeEventListener).toHaveBeenCalledWith(
      'mousemove',
      expect.any(Function)
    );
  });

  it('calls setMousePosition when mouse moves', () => {
    renderComponent();

    mockAddEventListener.mock.calls[0][1]({ clientX: 0, clientY: 1 });

    expect(mockSetMousePosition).toHaveBeenCalledWith({ x: 0, y: 1 });
  });
});
