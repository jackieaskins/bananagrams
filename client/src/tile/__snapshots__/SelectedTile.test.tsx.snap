// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectedTile /> does not render when no tile is selected 1`] = `""`;

exports[`<SelectedTile /> renders Tile when tile is selected 1`] = `
.emotion-01rduh4j-SelectedTile {
  position: fixed;
  left: 5px;
  top: 5px;
  z-index: 1000000;
}

<div
  className="emotion-0"
>
  <Memo(Tile)
    id="A1"
    letter="A"
  />
</div>
`;
